packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,168 @@
+# Data Hierarchy
+
+## Problem Statement
+
+There are scenarios where clients want to provide collapsible hierarchy to the rows being presented in the table, where both parent and children rows provide values in typically at least a subset of the same columns.
+
+For example, consider a table of geographic information that included location, population, flag, and possibly other location-related information. The location column could be either a continent, country, state, or any sub-regional construct. It would be reasonable to want to present this information in a way that allowed a user to collapse/expand any outer regional grouping, which itself would contain all rows whose locations existed inside it (e.g. 'United States' would be parent of all rows whose location was a U.S. state).
+
+### Lazy Loading
+
+In addition to just supporting data hierarchy with a fully provided set of data, we also need to support the scenario where a row is presented as a parent, but none of its children are currently represented in the data. Upon expanding the row, the client is then expected to load all of the data for that parent and provide it to the table (via the standard `setData` method).
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#890: Design for hierarchical data support in tables](https://github.com/ni/nimble/issues/890)
+-   [#861: Table hierarchical data support](https://github.com/ni/nimble/issues/861)
+-   [Prototype using flat list that includes a parentId](https://github.com/ni/nimble/tree/data-hierarchy-flat-list-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-qzwoshcidz.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+-   [Prototype with user data being hierarchical](https://github.com/ni/nimble/tree/data-hierarchy-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-yncupvnoes.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+
+## Implementation / Design
+
+Users will provide hierarchical data as a flat list (just as they do for non-hierarchical data), but will have to include a field in their data that is representative of the "parent Id" for that record of data.
+
+### `Table` API
+
+```ts
+public Table() {
+    ...
+    // This attribute determines which field in the record specifies the id that is the
+    // parent row. If it is not set, the row has no parent.
+    @attr({ attribute: 'parent-id-field-name' })
+    public parentIdFieldName?: string;
+
+    // This field would be an option for how we allow a user to specify the field
+    // name where they will provide a value indicating whether or not that row
+    // is a parent row. See the below discussion on lazy-loading. If it is not set,
+    // the row is presumed child-less (and thus will show no expand/collapse button).
```

Comment:
I can imagine a lazy loading scenario where we know the size of things in the hierarchy but don't want to load the full data / leaf nodes. i.e. I know I have 40 children. Is that a scenario we could expect today / should be able to accommodate?

Reply:
I guess I'm not fully understanding this scenario. Why wouldn't the proposed API handle this?

Reply:
Was hypothesizing about presenting that to user in a way they can see the count of children before expanding, or for having selection counts on lazy loaded sections be accurate

Reply:
My understanding is that parents would not show child counts. Additionally, group rows that had hierarchical data under them would only show counts for top-level entities. @leslieab, can you provide checks on me here? If I am right, does this satisfy your concern @rajsite?

Reply:
Yes, that's correct. We can revisit it, but that's the decision we landed on.

Reply:
After discussing offline I think my questions are resolved here. But there were follow-on questions around selection mode I captured in a separate thread. I think this thread is resolved.